src/controllers/ScrollRestorationController.ts

```diff
@@ -0,0 +1,82 @@
+import { Assert } from '../misc/Assert';
+import { ResultListEvents, IDisplayedNewResultsEventArgs } from '../events/ResultListEvents';
+import { QueryStateModel } from '../models/QueryStateModel';
+import { $$ } from '../utils/Dom';
+import { Logger } from '../misc/Logger';
+
+/**
+ * This class is instantiated on the root of {@link SearchInterface}
+ * if the {@link SearchInterface.options.enableScrollRestoration} option is set to true.
+ *
+ * Its job is to wait until the search results are displayed and the page height is long enough to
+ * restore to the last scroll position. It'll wait for up to `restorationTimeOutInMs` milliseconds for the page to reach its height.
+ *
+ * This is useful when back and forth navigation between the search interface and results is needed on the same tab.
+ *
+ */
+
+export class ScrollRestorationController {
+  public static stateAttributeName: string = 'yoffset';
+  /**
+   * Specifies the max amount of time to wait to restore the last scroll position.
+   */
+  public restorationTimeOutInMs: number = 5000;
+
+  private defaultYOffset: number = 0;
+  private tryToScrollIntervalInMs: number = 50;
+  private timeoutHandle = null;
+
+  constructor(public element: HTMLElement, public window: Window, public queryStateModel: QueryStateModel) {
+    Assert.exists(this.queryStateModel);
+
+    this.initStateAttribute();
+
+    this.window.addEventListener('beforeunload', () => this.saveScrollPositionInState());
+    $$(this.element).on(ResultListEvents.newResultsDisplayed, (event: Event, args: IDisplayedNewResultsEventArgs) =>
+      this.handleNewResultsDisplayed(args)
+    );
+  }
+
+  private initStateAttribute() {
+    this.queryStateModel.registerNewAttribute(ScrollRestorationController.stateAttributeName, this.defaultYOffset);
+  }
+
+  private saveScrollPositionInState() {
+    this.queryStateModel.set(ScrollRestorationController.stateAttributeName, this.window.pageYOffset);
```

Comment:
In some use case, clients will disable the URL management done by the UI, mostly because they are integrating into a website which already does client side URL manipulation for their own routing.

I'm wondering if we should not instead store this info in `localStorage` ? There's already a `LocalStorageUtils` class you could reuse.

And that would also mean you would not need any reference to `queryStateModel` passed in the constructor

Reply:
Yes it looks kind of weird to constantly update the QSM with the scroll position, it's also not something you want to pass over to someone if you send them a link.

Reply:
If using local Storage instead of the QSM, make sure you have some kind of expiration time. You don't want a user going to the same search page for unrelated reasons to be scrolled back down where he was in a previous session.

Reply:
Thinking back, my comment about query state model being disabled does not make much sense, since then the whole feature makes no sense (since we cannot know if you are performing the same exact query you were just doing before leaving the page, or if it's a new one, and then scrolling makes no sense). So forget about that part.

But, it is true what JF said, that having the URL constantly update as you scroll is a very weird. If we use local storage, then we need to make sure that we keep the "query state" also stored in some way with the scroll position, so that we only scroll restore if the URL parameters only matches what is stored. Some kind of hash function, or a simple JSON.stringify of the state.

Reply:
Well, in fact the URL just changes two times: on `beforeunload` and when the user goes back. So it's almost no perceptible. There's no easy way to copy that URL because the parameter gets removed from the URL right after the user goes back. The polyfill we saw about scroll restoration was using this URL approach, so it didn't seem that weird to me. But I could give local storage a try.

Reply:
In fact, SessionStorage would be ideal for this use case as it automatically expires not only on browser closed, but also on tab closed. No manual handling of expirations.